Updated units.py to include "week"

pdworz opened the pull request:
Note: This is my first time using git pull request. Please let me know if I need any changes

Commit:
Updated units.py to include "week"

Note: This is my first time using git. Please pull it so I know it works.

moorepants commented:
Please add a small test here:

https://github.com/sympy/sympy/blob/master/sympy/physics/tests/test_units.py

that involves a week in its calculation. Maybe how many weeks in a year?

Commit:
Updated test_units.py with a test for "week"

Note: this is a test for this pull request: https://github.com/sympy/sympy/pull/9206

pdworz commented:
OK, I made a test involving weeks in a year. (BTW thank you for helping me with my first ever Pull Request)

moorepants commented on sympy/physics/tests/test_units.py:
```diff
@@ -20,6 +20,8 @@ def test_units():
     assert integrate(t*m/s, (t, 1*s, 5*s)) == 12*m*s
     assert (t * m/s).integrate((t, 1*s, 5*s)) == 12*m*s
+    
+    assert str((year/week).n(2)) == '52.'
```
Why convert to a string here?

Does `assert year / week == 52` not work?

pdworz replied:
Good question. No, a year is not exactly 52 weeks. Some relevant output:

>>> year / week
36524219/700000
>>> (year / week).evalf()
52.1774557142857

moorepants replied:
Ok, that's fine, but I don't think the conversion to a string is a good way to check this. You can compare two floats like `assert abs(answer - expected) < tol` (for comparisons to non-zero numbers).

pdworz replied:
Yes, your suggestion is also how the tutorials recommend comparing expressions.

I made a change similar to your suggestion.

Commit:
Update test_units.py

moorepants commented on sympy/physics/tests/test_units.py:
```diff
@@ -20,6 +20,8 @@ def test_units():
     assert integrate(t*m/s, (t, 1*s, 5*s)) == 12*m*s
     assert (t * m/s).integrate((t, 1*s, 5*s)) == 12*m*s
+    
+    abs((year / week).evalf() - 52) < 0.2
```
There is no `assert` statement and why don't you use the "exact" value and a tighter tolerance? Or you can `evalf` to the exact number of digits you want and compare to a sympified float.

pdworz replied:
I did evalf to 15 digits precision, then cast it to a Python float for comparison.

(Thank you for your good advice and your patience. I really hope this eventually gets accepted.)

Commit:
updated test_units.py test precision of "week"

moorepants commented on sympy/physics/tests/test_units.py:
```diff
@@ -20,6 +20,8 @@ def test_units():
     assert integrate(t*m/s, (t, 1*s, 5*s)) == 12*m*s
     assert (t * m/s).integrate((t, 1*s, 5*s)) == 12*m*s
+    
+    assert float( (year / week).evalf(15) ) == 52.17745571428571
```
If you compare Python floats you must use the `abs(act - exp) < tol`. But if you compare SymPy `Float`s, which are arbitrary precision, then you can compare all the digits: `(year / week).evalf(15) == S(52.17745571428571)`. (I think).

pdworz replied:
Yes, you are correct. Updated. Note that I also imported `Float`.

Thank you for teaching me about the subtleties of testing using SymPy and Python types.

Commit:
update test_units.py to test week with SymPy Float

moorepants commented:
Ok this looks fine. Once the Travis tests pass it can be merged.

toolforger commented on sympy/physics/tests/test_units.py:
```diff
@@ -20,6 +20,8 @@ def test_units():
     assert integrate(t*m/s, (t, 1*s, 5*s)) == 12*m*s
     assert (t * m/s).integrate((t, 1*s, 5*s)) == 12*m*s
+    
+    assert (year / week).evalf(15) == Float(52.17745571428571)
```
Sorry for chiming in about this so late, and for confusing you by telling you the exact opposite of what other authorities have said, but I doubt that this test serves any useful purpose:
It does not allow easy manual verification that the test is correct. (In fact the test is harder to verify than the code that it tests, and that is the opposite of what a unit test should do. In fact we just saw that effect because we stumbled over ```year``` not being an integral multiple of ```day```.)
It does not fix any regression that ever happened.
It does not exercize any particular code path that is assumed to be prone to bugs.

I'd simply leave that test out.

moorepants replied:
Even if this test is as you say, this served the purpose of teaching someone how to write unit test. It can also possibly catch regressions and it adds virtually no time to the testing routines. How about we leave it in primarily for the sake of learning?

toolforger replied:
It indeed does not add noticeable time, but (a) it's a precedent for the bad practice of routinely adding test cases that do not really test anything, (b) the Float constructor is given a Python float so the argument that comparing with a Float is precise falls down (and may indeed fail on architectures with different IEEE architecture such as ARM, or faulty/configuration-dependent IEEE like i386), (c) if the test fails, and it's not something in the definition of ```week``` (as is most likely), the test failure is going to be a red herring for the maintainer.

It's of minimal value against regressions I think; any regression should be caught by a unit test that was written when the potentially recurring bug was fixed.

This does indeed teach the technicalities of adding a unit test, and as such, it does server its purpose.
It fails at teaching how to write *useful* unit tests though. And it's introducing problems in the test codebase where there was none before (the unexplained float literal, and the architecture-dependent float->Float conversion).

moorepants replied:
Please submit a PR against this branch that fixes everything then.

moorepants replied:
Also I'm well aware of the floating point comparisons. If you looked at the history of the thread you would see that I suggested a correct way to do that.

Secondly, all this PR then needs is quotes around the float value: `Float('52.17745571428571')` and there is no issue. I missed that and notice that I wrote "I think" when I suggested it.

Thanks for finding the error.

moorepants replied:
> It's of minimal value against regressions I think; any regression should be caught by a unit test that was written when the potentially recurring bug was fixed.

Regression tests are not only written when bugs are found. You write regression tests to catch future bugs regardless if you know about the bug or not. It  is especially useful to allow easy refactoring. Whether it is in the immediately tested function or something that is deep down the stack.

BTW, I'm annoyed that I'm arguing back and forth with you about a one line code change. Is this really that big of a deal to you? I'm bowing out of this discussion. You can take it from here.

toolforger replied:
Am 27.03.2015 um 23:40 schrieb Jason K. Moore:
> Regression tests are not only written when bugs are found. You write
> regression tests to catch future bugs regardless if you know about
> the bug or not.

Ah ok. I'd consider that a "correctness test" because the definition of 
"regression test" is another one.

 > It  is especially useful to allow easy refactoring.
> Whether it is in the immediately tested function or something that is
> deep down the stack.

I agree that correctness tests are useful, but I definitely disagree 
that we should add tests for something deep down the stack.

> BTW, I'm annoyed that I'm arguing back and forth with you about a one
> line code change. Is this really that big of a deal to you?

It's about teaching how to write unit tests.
I'm not concerned about the line of code as such.
I'm just annoyed by yet another overspecific test being added. This kind 
of stuff just cost me two weeks of camel shaving (and being unsure 
whether I'm changing semantics or not, because the tests lie to me and 
say the change is breaking things, while the Exprs that now come out are 
*almost* equivalent and I have to find out for each and every one 
whether the differences are something relevant or not).

toolforger replied:
Am 27.03.2015 um 23:36 schrieb Jason K. Moore:
> Secondly, all this PR then needs is quotes around the float value:
> `Float('52.17745571428571')` and there is no issue.

The issue actually stays. This just moves the floating-point conversion 
from the Python compiler to the Float constructor.

If a test is supposed to be in, I'd suggest testing against an exact value.
Looking at the definition of ```year``` (that's a few lines below 
```week```), it is defined as ```tropical_year``` which is 
```Rational('365.24219')*day```. That would be exact because 
```Rational('365.24219')``` isn't going through a floating-point conversion.
Using ```common_year``` would be even easier, that's just ```365``` and 
the test would not depend on any conversion considerations.

Or we could just compare against, say, ```day * 7```.
That would be silly, but it would test exactly what the definition of 
```week``` is supposed to be.

debugger22 commented:
Perhaps, it should be rebased first. There are 5 commits with almost same message.

@pdworz You can squash them with following command

```
$ git rebase -i HEAD~5
```
This will open a editor.
Replace `pick` with `s` of 4 commits(not the latest one) and save the file.
Another editor will open.
Comment out the message lines you don't want and save the file.
Now push it to your branch.
```
$ git push -f <your_remote_name> pdworz-patch-units-week
```

moorepants commented:
@debugger22 We typically try encourage avoiding rebasing, especially for beginners. There are two reasons: 1) for larger PRs that many people are working on, you change this history on people who've cloned, which isn't nice and 2) when you have to update to master and there are merge conflicts, rebasing is much more painful than merging and having a merge commit.

It doesn't really matter for this small PR but may for others.

See: https://github.com/sympy/sympy/wiki/Development-workflow#rebasing

> Note: merging is recommended over rebasing.

moorepants commented:
@debugger22 Here is a relevant email on rebasing vs merging: https://groups.google.com/forum/#!searchin/sympy/Recommending$20merging$20instead$20of$20rebasing/sympy/MnlVfmYYMio/rRt0ukqfFVIJ

debugger22 commented:
I think you are talking about "merging a branch with another branch" over "rebasing a branch with another branch".

What I suggested is just squashing the commits because of bad commit message. Since he's a beginner, I provided full method.

I'm ok anyway.

moorepants commented:
@debugger22 Yes you are correct there is a difference. I just try to avoid bringing up rebasing for newcomers so they don't get into the bad habit of rewriting the git history from the get go.

moorepants commented:
@pdworz Please squash the commits as to @debugger22's suggestion.

pdworz commented:
I have created a mess. I made an initial change on my computer, and then used the github web interface to make changes. The web interface does not support squashing.

I am sorry for this mess I have created. Thank you for putting up with my growing pains.

This pull request is important because it is in partial fulfillment of the requirements of the Google Summer of Code application. Can you please tell me what I can do to get it merged? Thank you for your time and your patience.

moorepants commented:
Squashing the commits is not a necessity. But for future pull requests be sure to write informative commit messages, see:

https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages

Secondly, it may be possible to edit files through the Github interface but I would not recommend it. It is impossible to debug code and run the tests yourself when doing so. The Github interface will also lead to many useless commits as you try to "debug"  your code.

If you haven't read through the development workflow page, please read the whole thing:

https://github.com/sympy/sympy/wiki/Development-workflow

And if you want to squash the commit messages, it is possible. But best to clone the code locally first.

pdworz commented:
Yes, you are correct on all counts. I should write better commit messages, with bodies that include explanations and a code sample. Now I know that the git program is more powerful than the web interface. I had a problem with the tests: they took very long on my netbook, so I should develop on a workstation. I read/skimmed Development Workflow, but it has a lot of information, and I need more than one day to learn it all.

I am very lucky that you have dedicated so much time to me. I will always remember this experience.

debugger22 commented:
>  I had a problem with the tests: they took very long on my netbook, so I should develop on a workstation

You can also run partial tests on your system if you are sure what you are doing.

If you want to run tests in the directory `sympy/physics/tests/`
```
$ ./bin/test sympy/physics/tests/
```
You can also run tests of one file by passing its path
```
$ ./bin/test sympy/physics/tests/test_units.py
```
You can also run a single test function by passing its name
```
$ ./bin/test -k test_units()
```

Commit:
Update test_units.py

Commit:
Update test_units.py - removed "Float"

The `from sympy import ... Float` remained from earlier changes. It is no longer needed.

pdworz commented:
Hi @debugger22. You are 100% correct, and I did run tests for just physics, but I was not convinced that this is enough; I was afraid that the full test is needed every time.

Anyway, I got rid of the test, and cleaned up the patch to only one line in one file. This is my first ever patch, so I am sorry that I have made so many clumsy commits. I am embarrassed that I triggered Travis so many times. I have learned many lessons. I am lucky that the SymPy community is so helpful, I want to be a part of this community.

Can someone please consider pulling this patch? I know that it is trivial, but it is part of the requirements for the Google Summer of Code application. Please help a new hopeful SymPy developer :)

debugger22 commented:
+1

certik commented:
@pdworz what is the status of this? Currently the PR has only a one line change without a test. You should add a test for this.

toolforger commented:
How would a meaningful test look like?
The only thing I can imagine is the mere existence of `week`. Everything I can imagine beyond that would either be a more-or-less verbatim copy of that line, or test things that aren't properties of that line (such as evaluating the number of seconds in a week, which would mainly test the number of seconds per day and the validity of the days-to-seconds conversion).

certik commented:
See the tests here: https://github.com/pdworz/sympy/blob/3250b188ad457deeea063e7aca4aed71bef95eb6/sympy/physics/tests/test_units.py, it should test that `week = 7*days` and not `week = 8*days`.

toolforger commented:
I doubt the value of tests that just repeat the definition, but if that's what it takes...

certik commented:
It's a one line test. It tests that it can be imported properly, that it works.
